<img align="left" src="https://ithaka-labs.s3.amazonaws.com/static-files/images/tdm/tdmdocs/tapi-logo-small.png" />

This notebook free for educational reuse under [Creative Commons CC BY License](https://creativecommons.org/licenses/by/4.0/).

Created by [Firstname Lastname](https://) for the 2022 Text Analysis Pedagogy Institute, with support from the [National Endowment for the Humanities](https://neh.gov), [JSTOR Labs](https://labs.jstor.org/), and [University of Arizona Libraries](https://new.library.arizona.edu/).

For questions/comments/improvements, email author@email.address.<br />
____

# `Introduction to Pandas` `2`

This is lesson `1` of 3 in the educational series on `TOPIC`. This notebook is intended `to teach XXX and introduce the concepts of XXXX`. 

**Audience:** `Teachers` / `Learners` / `Researchers`

**Use case:** `Tutorial` / `How-To` / `Reference` / `Explanation` 

`Include the use case definition from [here](https://constellate.org/docs/documentation-categories)`

**Difficulty:** `Beginner` / `Intermediate` / `Advanced`

`Beginner assumes users are relatively new to Python and Jupyter Notebooks. The user is helped step-by-step with lots of explanatory text.`
`Intermediate assumes users are familiar with Python and have been programming for 6+ months. Code makes up a larger part of the notebook and basic concepts related to Python are not explained.`
`Advanced assumes users are very familiar with Python and have been programming for years, but they may not be familiar with the process being explained.`

**Completion time:** `90 minutes`

**Knowledge Required:** 
```
* Python basics (variables, flow control, functions, lists, dictionaries)
* Object-oriented programming (classes, instances, inheritance)

These should be general skills but can mention a particular library
```

**Knowledge Recommended:**
```
* Basic file operations (open, close, read, write)
```

**Learning Objectives:**
After this lesson, learners will be able to:
```
1. How to Find Data in a DataFrame
2. How to Organize Data
3. How to Clean Data
```
___

# Required Python Libraries
`List out any libraries used and what they are used for`
* Pandas

## Install Required Libraries

In [ ]:
### Install Libraries ###

# Using !pip installs
!pip install pandas

# Introduction

```
Pandas is a Python library in Python that allows for you to easily work with tabular data that is often stored in Excel files or .csv files. It is often considered one of the most vital Python libraries for data analysis because of its ease of use. There are other Python libraries that allow you to work with Excel data, such as XLSX and XLRD, which are leveraged by Pandas on the back end to interact with Excel spreadsheets. However, most tabular data is stored in .csv files, or Comma Separated Values files. There are other variations of this same structure (such as .tsv), but they all work the same. They use a special character to denote movement in the table to the next cell.
```

# Finding Data in DataFrame

## Covered in this Section

1) How to Find Column Names<br>
2) How to Get a Quick Sense of the Dataset<br>
3) How to Grab a Specific Range of Rows<br>
4) How to Get a Quick Quantitative Understanding of the Dataset<br>
5) How to Find Specific Information in the Dataset<br>
6) How to use Or in a DataFrame query<br>

## Video

In [1]:
%%HTML
<center>
<iframe width="560" height="315" src="https://www.youtube.com/embed/IFVkOCnH_Qs" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
</center>

## About the Titanic Dataset

In [2]:
import pandas as pd

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/f/fd/RMS_Titanic_3.jpg/1200px-RMS_Titanic_3.jpg">

In the next two sections, we will be using the same dataset, the infamous *Titanic Survivor* dataset that was put out by Kaggle a few years ago. It has since become a staple dataset in machine learning and data science communities because of the breadth of data that it offers on the Titanic passengers. The goal of most machine learning challenges is to use this dataset to train a model that can accurately predict if a passenger would have survived given certain gender, economic, and class conditions.

In these notebooks, we will be using a cleaned version of the dataset available on the GitHub page below. We will be using it because of its breadth of its qualitative and quantitative data.

In this notebook, we are strictly interested in finding key data in the DataFrame. By the end of the notebook, you should have a good understanding of why working with CSV data in Python via Pandas is far more powerful and easy than using Excel.

In [5]:
#data obtained from => https://github.com/datasciencedojo/datasets , but originally from Kaggle
df = pd.read_csv("data/titanic.csv")

## How to Find Column Data

Let's presume that we just obtained this dataset. Let's also presume we know nothing about it. Our first job is to get a sense of the data. Maybe we want to know all the columns that the dataset contains. To find this information, we can use df.columns

In [3]:
df.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

With our knowledge from the last notebook, we now know how to convert this into a list!

In [4]:
cols = df.columns.tolist()

In [5]:
cols

['PassengerId',
 'Survived',
 'Pclass',
 'Name',
 'Sex',
 'Age',
 'SibSp',
 'Parch',
 'Ticket',
 'Fare',
 'Cabin',
 'Embarked']

## How to Get a Quick Sense of the Dataset with df.head()

Another way to get a quick sense of the data is to use df.head(). This allows us to return to top pieces of information from the dataset. By default, head returns the top 5 rows.

In [6]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

We can pass in an argument to return more than 5. Let's try and display 20.

In [7]:
df.head(20)

PassengerId  Survived  Pclass  \
0             1         0       3   
1             2         1       1   
2             3         1       3   
3             4         1       1   
4             5         0       3   
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   
12           13         0       3   
13           14         0       3   
14           15         0       3   
15           16         1       2   
16           17         0       3   
17           18         1       2   
18           19         0       3   
19           20         1       3   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                              Heikkinen, Miss. Laina  female  26.0      0   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                            Allen, Mr. William Henry    male  35.0      0   
5                                    Moran, Mr. James    male   NaN      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
12                     Saundercock, Mr. William Henry    male  20.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
17                       Williams, Mr. Charles Eugene    male   NaN      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                            Masselmani, Mrs. Fatima  female   NaN      0   

    Parch            Ticket     Fare Cabin Embarked  
0       0         A/5 21171   7.2500   NaN        S  
1       0          PC 17599  71.2833   C85        C  
2       0  STON/O2. 3101282   7.9250   NaN        S  
3       0            113803  53.1000  C123        S  
4       0            373450   8.0500   NaN        S  
5       0            330877   8.4583   NaN        Q  
6       0             17463  51.8625   E46        S  
7       1            349909  21.0750   NaN        S  
8       2            347742  11.1333   NaN        S  
9       0            237736  30.0708   NaN        C  
10      1           PP 9549  16.7000    G6        S  
11      0            113783  26.5500  C103        S  
12      0         A/5. 2151   8.0500   NaN        S  
13      5            347082  31.2750   NaN        S  
14      0            350406   7.8542   NaN        S  
15      0            248706  16.0000   NaN        S  
16      1            382652  29.1250   NaN        Q  
17      0            244373  13.0000   NaN        S  
18      0            345763  18.0000   NaN        S  
19      0              2649   7.2250   NaN        C

## How to Grab a Specific Range of Rows with df.iloc[]

Sometimes, however, you need to grab a specific range of rows. Let's say I am interested in rows 5-20. To grab this data, we can use the df.iloc[] command that we met in the last notebook. This will allow us to pass a specific index range like a list.

In [8]:
df.iloc[5:20]

PassengerId  Survived  Pclass  \
5             6         0       3   
6             7         0       1   
7             8         0       3   
8             9         1       3   
9            10         1       2   
10           11         1       3   
11           12         1       1   
12           13         0       3   
13           14         0       3   
14           15         0       3   
15           16         1       2   
16           17         0       3   
17           18         1       2   
18           19         0       3   
19           20         1       3   

                                                 Name     Sex   Age  SibSp  \
5                                    Moran, Mr. James    male   NaN      0   
6                             McCarthy, Mr. Timothy J    male  54.0      0   
7                      Palsson, Master. Gosta Leonard    male   2.0      3   
8   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                 Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                    Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
12                     Saundercock, Mr. William Henry    male  20.0      0   
13                        Andersson, Mr. Anders Johan    male  39.0      1   
14               Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
15                   Hewlett, Mrs. (Mary D Kingcome)   female  55.0      0   
16                               Rice, Master. Eugene    male   2.0      4   
17                       Williams, Mr. Charles Eugene    male   NaN      0   
18  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
19                            Masselmani, Mrs. Fatima  female   NaN      0   

    Parch     Ticket     Fare Cabin Embarked  
5       0     330877   8.4583   NaN        Q  
6       0      17463  51.8625   E46        S  
7       1     349909  21.0750   NaN        S  
8       2     347742  11.1333   NaN        S  
9       0     237736  30.0708   NaN        C  
10      1    PP 9549  16.7000    G6        S  
11      0     113783  26.5500  C103        S  
12      0  A/5. 2151   8.0500   NaN        S  
13      5     347082  31.2750   NaN        S  
14      0     350406   7.8542   NaN        S  
15      0     248706  16.0000   NaN        S  
16      1     382652  29.1250   NaN        Q  
17      0     244373  13.0000   NaN        S  
18      0     345763  18.0000   NaN        S  
19      0       2649   7.2250   NaN        C

## How to Get a Quick Quantitative Understanding of the Dataset with Describe()

When working with numerical, quantitative data, we can automatically grab all numerical rows and learn a lot about the data with df.describe(). This will return the count, mean, standard deviation, minimum, maximum, etc. of our quantitative data. Sometimes, this data is useful to compute in such a way, such as the column Survived. Here, we can see that roughly .38 or 38% of the passengers (in this dataset) survived. Other numerical data does not really lend itself well to this kind of analysis, e.g. PassengerId which a unique numerical number corresponding to each passenger.

In [9]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

## How to Find Specific Information in the Dataset with df.loc

As researchers, however, we often need to find targeted information in our dataset. Let's say I am researching female passengers on the Titanic. I can achieve this fairly easily in Excel and in Python with Pandas. To do this in Pandas, you can use df.loc[] to pass a specific argument. In the example below, we are stating that we are looking for the columns in the DataFrame of "Sex" that match the string "female". Try to recreate this and find all males in the dataset on your own. Note at the bottom of the DataFrame 314x12. This is the dimensions of the DataFrame. We can see that we have 314 results. 

In [10]:
df.loc[df["Sex"] == "female"]

PassengerId  Survived  Pclass  \
1              2         1       1   
2              3         1       3   
3              4         1       1   
8              9         1       3   
9             10         1       2   
..           ...       ...     ...   
880          881         1       2   
882          883         0       3   
885          886         0       3   
887          888         1       1   
888          889         0       3   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
..                                                 ...     ...   ...    ...   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
882                       Dahlberg, Miss. Gerda Ulrika  female  22.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   

     Parch            Ticket     Fare Cabin Embarked  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
8        2            347742  11.1333   NaN        S  
9        0            237736  30.0708   NaN        C  
..     ...               ...      ...   ...      ...  
880      1            230433  26.0000   NaN        S  
882      0              7552  10.5167   NaN        S  
885      5            382652  29.1250   NaN        Q  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  

[314 rows x 12 columns]

But that's not all df.loc[] can do. We can add multiple arguments to our search query here, rather like SQL. In the example below, note the addition of the parentheses around df\["Sex"\] == "female" this denotes one parameter of our search. We then use the & to add a second parameter. In this case, we are interested in not only returning those that are female, but also those who are in Pclass, or Passenger Class 1, i.e. First Class. When we execute the command below, we see that we have found 94 results.

In [11]:
df.loc[(df["Sex"] == "female") & (df["Pclass"] == 1)]

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
11            12         1       1   
31            32         1       1   
52            53         1       1   
..           ...       ...     ...   
856          857         1       1   
862          863         1       1   
871          872         1       1   
879          880         1       1   
887          888         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
31      Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN      1   
52            Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0      1   
..                                                 ...     ...   ...    ...   
856         Wick, Mrs. George Dennick (Mary Hitchcock)  female  45.0      1   
862  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  female  48.0      0   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   

     Parch    Ticket      Fare Cabin Embarked  
1        0  PC 17599   71.2833   C85        C  
3        0    113803   53.1000  C123        S  
11       0    113783   26.5500  C103        S  
31       0  PC 17569  146.5208   B78        C  
52       0  PC 17572   76.7292   D33        C  
..     ...       ...       ...   ...      ...  
856      1     36928  164.8667   NaN        S  
862      0     17466   25.9292   D17        S  
871      1     11751   52.5542   D35        S  
879      1     11767   83.1583   C50        C  
887      0    112053   30.0000   B42        S  

[94 rows x 12 columns]

Let's say now we are interested in only finding those that did not survive of the 94. To do that, we would simply add a third argument to our query. We ask Pandas to only show those whose Survived value is 0, or False. In other words, we are seeking to find those who did not survive.

In [12]:
df.loc[(df["Sex"] == "female") & (df["Pclass"] == 1) & (df["Survived"] == 0)]

PassengerId  Survived  Pclass  \
177          178         0       1   
297          298         0       1   
498          499         0       1   

                                                Name     Sex   Age  SibSp  \
177                       Isham, Miss. Ann Elizabeth  female  50.0      0   
297                     Allison, Miss. Helen Loraine  female   2.0      1   
498  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.0      1   

     Parch    Ticket      Fare    Cabin Embarked  
177      0  PC 17595   28.7125      C49        C  
297      2    113781  151.5500  C22 C26        S  
498      2    113781  151.5500  C22 C26        S

With this data, I not only have qualitative information about each passenger, I also have some significant quantitative information as well. I know that of the 314 people in the dataset with the sex of female, 94 were in first class and of those 94, all survived except 3. Let's now pose a question and I think you may already know the answer. Did class in any way play a role in the chance off survival on the Titanic. We don't need a fancy machine learning algorithm to help us answer this question. We can simply examine the results from our query. Let's find the total number of passengers that are identified as female, but not in first class, and then find the number of those who did not survive.

In [13]:
df.loc[(df["Sex"] == "female") & (df["Pclass"] > 1)]

PassengerId  Survived  Pclass  \
2              3         1       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
14            15         0       3   
..           ...       ...     ...   
875          876         1       3   
880          881         1       2   
882          883         0       3   
885          886         0       3   
888          889         0       3   

                                                  Name     Sex   Age  SibSp  \
2                               Heikkinen, Miss. Laina  female  26.0      0   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
..                                                 ...     ...   ...    ...   
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
882                       Dahlberg, Miss. Gerda Ulrika  female  22.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   

     Parch            Ticket     Fare Cabin Embarked  
2        0  STON/O2. 3101282   7.9250   NaN        S  
8        2            347742  11.1333   NaN        S  
9        0            237736  30.0708   NaN        C  
10       1           PP 9549  16.7000    G6        S  
14       0            350406   7.8542   NaN        S  
..     ...               ...      ...   ...      ...  
875      0              2667   7.2250   NaN        C  
880      1            230433  26.0000   NaN        S  
882      0              7552  10.5167   NaN        S  
885      5            382652  29.1250   NaN        Q  
888      2        W./C. 6607  23.4500   NaN        S  

[220 rows x 12 columns]

In [14]:
df.loc[(df["Sex"] == "female") & (df["Pclass"] > 1) & (df["Survived"] == 0)]

PassengerId  Survived  Pclass  \
14            15         0       3   
18            19         0       3   
24            25         0       3   
38            39         0       3   
40            41         0       3   
..           ...       ...     ...   
854          855         0       2   
863          864         0       3   
882          883         0       3   
885          886         0       3   
888          889         0       3   

                                                  Name     Sex   Age  SibSp  \
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
18   Vander Planke, Mrs. Julius (Emelia Maria Vande...  female  31.0      1   
24                       Palsson, Miss. Torborg Danira  female   8.0      3   
38                  Vander Planke, Miss. Augusta Maria  female  18.0      2   
40      Ahlin, Mrs. Johan (Johanna Persdotter Larsson)  female  40.0      1   
..                                                 ...     ...   ...    ...   
854      Carter, Mrs. Ernest Courtenay (Lilian Hughes)  female  44.0      1   
863                  Sage, Miss. Dorothy Edith "Dolly"  female   NaN      8   
882                       Dahlberg, Miss. Gerda Ulrika  female  22.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   

     Parch      Ticket     Fare Cabin Embarked  
14       0      350406   7.8542   NaN        S  
18       0      345763  18.0000   NaN        S  
24       1      349909  21.0750   NaN        S  
38       0      345764  18.0000   NaN        S  
40       0        7546   9.4750   NaN        S  
..     ...         ...      ...   ...      ...  
854      0      244252  26.0000   NaN        S  
863      2    CA. 2343  69.5500   NaN        S  
882      0        7552  10.5167   NaN        S  
885      5      382652  29.1250   NaN        Q  
888      2  W./C. 6607  23.4500   NaN        S  

[78 rows x 12 columns]

As we can see, we have 220 total people identified as female with 78 not surviving. We can perform some quick calculations to understand better our results.

In [15]:
x = 3/94
x

0.031914893617021274

In [16]:
y = 78/220
y

0.35454545454545455

In this case, x is the death rate for a first class female, whereas y is the death rate for those not in first class and female. The result is clear. The chance for survival was significantly greater for those identified as female if they were in first class, roughly 32.5% difference. Were there other factors at play, possibly. But that is beyond the point here. I am simply trying to illustrate how to obtain data to begin framing the research question at hand.

Although we can achieve this same analysis in Excel, doing this in Python allows us to leverage the power of an entire programming language with the data at hand. This is a simple example of the power of Pandas over Excel. As we move forward in the next notebooks, you will see more examples of this.

## How to Query with "OR" (|) on a DataFrame

Above we saw how to use df.loc to structure logical arguments for finding two specific conditions with "&". This indicated that both conditions must be true to be returned as a result. In case you don't remember it, it looked like this:

In [17]:
df.loc[(df["Sex"] == "female") & (df["Pclass"] > 1)]

PassengerId  Survived  Pclass  \
2              3         1       3   
8              9         1       3   
9             10         1       2   
10            11         1       3   
14            15         0       3   
..           ...       ...     ...   
875          876         1       3   
880          881         1       2   
882          883         0       3   
885          886         0       3   
888          889         0       3   

                                                  Name     Sex   Age  SibSp  \
2                               Heikkinen, Miss. Laina  female  26.0      0   
8    Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
14                Vestrom, Miss. Hulda Amanda Adolfina  female  14.0      0   
..                                                 ...     ...   ...    ...   
875                   Najib, Miss. Adele Kiamie "Jane"  female  15.0      0   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
882                       Dahlberg, Miss. Gerda Ulrika  female  22.0      0   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   

     Parch            Ticket     Fare Cabin Embarked  
2        0  STON/O2. 3101282   7.9250   NaN        S  
8        2            347742  11.1333   NaN        S  
9        0            237736  30.0708   NaN        C  
10       1           PP 9549  16.7000    G6        S  
14       0            350406   7.8542   NaN        S  
..     ...               ...      ...   ...      ...  
875      0              2667   7.2250   NaN        C  
880      1            230433  26.0000   NaN        S  
882      0              7552  10.5167   NaN        S  
885      5            382652  29.1250   NaN        Q  
888      2        W./C. 6607  23.4500   NaN        S  

[220 rows x 12 columns]

We can do the same conditional approach, but rather than using and, we can also tell Pandas "Or" with the "|", that vertical character you probably never use located above the "Enter" button on a standard American keyboard. Let's try and grab all first and second class passengers. We can specify if the Pclass is either 1 OR 2.

In [18]:
df.loc[(df["Pclass"] == 1) | (df["Pclass"] == 2)]

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
9             10         1       2   
11            12         1       1   
..           ...       ...     ...   
880          881         1       2   
883          884         0       2   
886          887         0       2   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
9                  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
880       Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0      0   
883                      Banfield, Mr. Frederick James    male  28.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch            Ticket     Fare Cabin Embarked  
1        0          PC 17599  71.2833   C85        C  
3        0            113803  53.1000  C123        S  
6        0             17463  51.8625   E46        S  
9        0            237736  30.0708   NaN        C  
11       0            113783  26.5500  C103        S  
..     ...               ...      ...   ...      ...  
880      1            230433  26.0000   NaN        S  
883      0  C.A./SOTON 34068  10.5000   NaN        S  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  

[400 rows x 12 columns]

# Organizing the DataFrame

## Covered in this Section

1) How to Sort Data by Single Column<br>
2) How to Reverse Sort Data by Single Column<br>
3) How to Sort Data by Multiple Columns<br>
4) How to Sort Data by Multiple Columns with Different Values Organized Differently<br>

## Video

In [3]:
%%HTML
<center>
<iframe width="560" height="315" src="https://www.youtube.com/embed/1G1ursLeZNg" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
</center>

## How to Sort Data By Single Column

Let's take a look at our DataFrame once again.

In [6]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In this scenario, I am interested in sorting the data (rather like Excel). Rather than using the sort feature in Excel, we can use the df.sort_values() method in Python. This will take one argument, specifically the column that you want to organize by. By default, this will be ascending. Let's do this by class. In other words, sort the DataFrame so that those who were in first class appear first and those in third class appear last. We will do this by passing the argument "Pclass", the column name corresponding to Passenger Class.

In [3]:
df.sort_values("Pclass")

PassengerId  Survived  Pclass  \
445          446         1       1   
310          311         1       1   
309          310         1       1   
307          308         1       1   
306          307         1       1   
..           ...       ...     ...   
379          380         0       3   
381          382         1       3   
382          383         0       3   
371          372         0       3   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
445                          Dodge, Master. Washington    male   4.0      0   
310                     Hays, Miss. Margaret Bechstein  female  24.0      0   
309                     Francatelli, Miss. Laura Mabel  female  30.0      0   
307  Penasco y Castellana, Mrs. Victor de Satode (M...  female  17.0      1   
306                            Fleming, Miss. Margaret  female   NaN      0   
..                                                 ...     ...   ...    ...   
379                        Gustafsson, Mr. Karl Gideon    male  19.0      0   
381                        Nakid, Miss. Maria ("Mary")  female   1.0      0   
382                                 Tikkanen, Mr. Juho    male  32.0      0   
371                          Wiklund, Mr. Jakob Alfred    male  18.0      1   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch             Ticket      Fare Cabin Embarked  
445      2              33638   81.8583   A34        S  
310      0              11767   83.1583   C54        C  
309      0           PC 17485   56.9292   E36        C  
307      0           PC 17758  108.9000   C65        C  
306      0              17421  110.8833   NaN        C  
..     ...                ...       ...   ...      ...  
379      0             347069    7.7750   NaN        S  
381      2               2653   15.7417   NaN        C  
382      0  STON/O 2. 3101293    7.9250   NaN        S  
371      0            3101267    6.4958   NaN        S  
890      0             370376    7.7500   NaN        Q  

[891 rows x 12 columns]

## How to Reverse Sort Data by Single Column

As we can see, our data is now appearing as expected. We can pass additional keyword arguments to sort the data in the opposite direction, or descending by setting ascending to False. See the example below.

In [4]:
df.sort_values("Pclass", ascending=False)

PassengerId  Survived  Pclass  \
0              1         0       3   
511          512         0       3   
500          501         0       3   
501          502         0       3   
502          503         0       3   
..           ...       ...     ...   
102          103         0       1   
710          711         1       1   
711          712         0       1   
712          713         1       1   
445          446         1       1   

                                                 Name     Sex   Age  SibSp  \
0                             Braund, Mr. Owen Harris    male  22.0      1   
511                                 Webber, Mr. James    male   NaN      0   
500                                  Calic, Mr. Petar    male  17.0      0   
501                               Canavan, Miss. Mary  female  21.0      0   
502                    O'Sullivan, Miss. Bridget Mary  female   NaN      0   
..                                                ...     ...   ...    ...   
102                         White, Mr. Richard Frasar    male  21.0      0   
710  Mayne, Mlle. Berthe Antonine ("Mrs de Villiers")  female  24.0      0   
711                                Klaber, Mr. Herman    male   NaN      0   
712                          Taylor, Mr. Elmer Zebley    male  48.0      1   
445                         Dodge, Master. Washington    male   4.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
511      0  SOTON/OQ 3101316   8.0500   NaN        S  
500      0            315086   8.6625   NaN        S  
501      0            364846   7.7500   NaN        Q  
502      0            330909   7.6292   NaN        Q  
..     ...               ...      ...   ...      ...  
102      1             35281  77.2875   D26        S  
710      0          PC 17482  49.5042   C90        C  
711      0            113028  26.5500  C124        S  
712      0             19996  52.0000  C126        S  
445      2             33638  81.8583   A34        S  

[891 rows x 12 columns]

## How to Sort Data by Multiple Columns

Again, we can see the power of Pandas over Excel by the simplicity of altering our command to include multiple columns. Let's say that we want to sort all the data by Pclass, then we want that data organized again by sex, so that all male and female passengers appear in order. We can do this by passing the argument of what we want organized as a list. Note the order of the list as well. The columns that appear earlier in the list correspond to those that receive primacy in the ascending. In other words, we organize by passenger class firsit, then sex. In this case, the method head, is simply showing the top 100 rows.

In [5]:
df.sort_values(["Pclass", "Sex"]).head(100)

PassengerId  Survived  Pclass  \
1             2         1       1   
3             4         1       1   
11           12         1       1   
31           32         1       1   
52           53         1       1   
..          ...       ...     ...   
23           24         1       1   
27           28         0       1   
30           31         0       1   
34           35         0       1   
35           36         0       1   

                                                 Name     Sex   Age  SibSp  \
1   Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3        Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
11                           Bonnell, Miss. Elizabeth  female  58.0      0   
31     Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN      1   
52           Harper, Mrs. Henry Sleeper (Myna Haxtun)  female  49.0      1   
..                                                ...     ...   ...    ...   
23                       Sloper, Mr. William Thompson    male  28.0      0   
27                     Fortune, Mr. Charles Alexander    male  19.0      3   
30                           Uruchurtu, Don. Manuel E    male  40.0      0   
34                            Meyer, Mr. Edgar Joseph    male  28.0      1   
35                     Holverson, Mr. Alexander Oskar    male  42.0      1   

    Parch    Ticket      Fare        Cabin Embarked  
1       0  PC 17599   71.2833          C85        C  
3       0    113803   53.1000         C123        S  
11      0    113783   26.5500         C103        S  
31      0  PC 17569  146.5208          B78        C  
52      0  PC 17572   76.7292          D33        C  
..    ...       ...       ...          ...      ...  
23      0    113788   35.5000           A6        S  
27      2     19950  263.0000  C23 C25 C27        S  
30      0  PC 17601   27.7208          NaN        C  
34      0  PC 17604   82.1708          NaN        C  
35      0    113789   52.0000          NaN        S  

[100 rows x 12 columns]

As with before, we can control how the data is sorted, either ascending or descending. If we set ascending to False, we organize all items in the list by this method. We can do this with the sample code below.

In [6]:
df.sort_values(["Pclass", "Sex"], ascending=False)

PassengerId  Survived  Pclass  \
0              1         0       3   
4              5         0       3   
5              6         0       3   
7              8         0       3   
12            13         0       3   
..           ...       ...     ...   
856          857         1       1   
862          863         1       1   
871          872         1       1   
879          880         1       1   
887          888         1       1   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
5                                     Moran, Mr. James    male   NaN      0   
7                       Palsson, Master. Gosta Leonard    male   2.0      3   
12                      Saundercock, Mr. William Henry    male  20.0      0   
..                                                 ...     ...   ...    ...   
856         Wick, Mrs. George Dennick (Mary Hitchcock)  female  45.0      1   
862  Swift, Mrs. Frederick Joel (Margaret Welles Ba...  female  48.0      0   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   

     Parch     Ticket      Fare Cabin Embarked  
0        0  A/5 21171    7.2500   NaN        S  
4        0     373450    8.0500   NaN        S  
5        0     330877    8.4583   NaN        Q  
7        1     349909   21.0750   NaN        S  
12       0  A/5. 2151    8.0500   NaN        S  
..     ...        ...       ...   ...      ...  
856      1      36928  164.8667   NaN        S  
862      0      17466   25.9292   D17        S  
871      1      11751   52.5542   D35        S  
879      1      11767   83.1583   C50        C  
887      0     112053   30.0000   B42        S  

[891 rows x 12 columns]

## How to Sort Data by Multiple Columns with Different Values Organized Differently

What if we want to organize the data differently. By this I mean, we want for all the data to be organized by passenger class first and for that data to be ascending (1, 2, 3), but we want the sex of the passengers to be organized descending (male, female, rather than female, male). To achieve this, we can pass a list to ascending with 0s and 1s. 0 is False and 1 is True.

In [7]:
df.sort_values(["Pclass", "Sex"], ascending=[1,0])

PassengerId  Survived  Pclass                                      Name  \
6              7         0       1                   McCarthy, Mr. Timothy J   
23            24         1       1              Sloper, Mr. William Thompson   
27            28         0       1            Fortune, Mr. Charles Alexander   
30            31         0       1                  Uruchurtu, Don. Manuel E   
34            35         0       1                   Meyer, Mr. Edgar Joseph   
..           ...       ...     ...                                       ...   
863          864         0       3         Sage, Miss. Dorothy Edith "Dolly"   
875          876         1       3          Najib, Miss. Adele Kiamie "Jane"   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   

        Sex   Age  SibSp  Parch      Ticket      Fare        Cabin Embarked  
6      male  54.0      0      0       17463   51.8625          E46        S  
23     male  28.0      0      0      113788   35.5000           A6        S  
27     male  19.0      3      2       19950  263.0000  C23 C25 C27        S  
30     male  40.0      0      0    PC 17601   27.7208          NaN        C  
34     male  28.0      1      0    PC 17604   82.1708          NaN        C  
..      ...   ...    ...    ...         ...       ...          ...      ...  
863  female   NaN      8      2    CA. 2343   69.5500          NaN        S  
875  female  15.0      0      0        2667    7.2250          NaN        C  
882  female  22.0      0      0        7552   10.5167          NaN        S  
885  female  39.0      0      5      382652   29.1250          NaN        Q  
888  female   NaN      1      2  W./C. 6607   23.4500          NaN        S  

[891 rows x 12 columns]

What is particularly nice about Pandas over Excel is that this operation scales nicely. If we want to add more methods of sorting, we can do that too by simple increasing the indices of our lists. Always make sure that the length of your lists match, however. In other words, do not have 3 attributes to sort by and 2 items in your ascending list. In this case, we want to organize by passenger class, sex, and age with passenger class ascending, sex descending, and age ascending. Let's see what that would look like.

In [8]:
df.sort_values(["Pclass", "Sex", "Age"], ascending=[1,0,1])

PassengerId  Survived  Pclass  \
305          306         1       1   
445          446         1       1   
802          803         1       1   
550          551         1       1   
505          506         0       1   
..           ...       ...     ...   
697          698         1       3   
727          728         1       3   
792          793         0       3   
863          864         0       3   
888          889         0       3   

                                           Name     Sex    Age  SibSp  Parch  \
305              Allison, Master. Hudson Trevor    male   0.92      1      2   
445                   Dodge, Master. Washington    male   4.00      0      2   
802         Carter, Master. William Thornton II    male  11.00      1      2   
550                 Thayer, Mr. John Borland Jr    male  17.00      0      2   
505  Penasco y Castellana, Mr. Victor de Satode    male  18.00      1      0   
..                                          ...     ...    ...    ...    ...   
697            Mullens, Miss. Katherine "Katie"  female    NaN      0      0   
727                    Mannion, Miss. Margareth  female    NaN      0      0   
792                     Sage, Miss. Stella Anna  female    NaN      8      2   
863           Sage, Miss. Dorothy Edith "Dolly"  female    NaN      8      2   
888    Johnston, Miss. Catherine Helen "Carrie"  female    NaN      1      2   

         Ticket      Fare    Cabin Embarked  
305      113781  151.5500  C22 C26        S  
445       33638   81.8583      A34        S  
802      113760  120.0000  B96 B98        S  
550       17421  110.8833      C70        C  
505    PC 17758  108.9000      C65        C  
..          ...       ...      ...      ...  
697       35852    7.7333      NaN        Q  
727       36866    7.7375      NaN        Q  
792    CA. 2343   69.5500      NaN        S  
863    CA. 2343   69.5500      NaN        S  
888  W./C. 6607   23.4500      NaN        S  

[891 rows x 12 columns]

As we move forward throughout these notebooks, we will explore more robust ways to sort and organize our data. For now, you should feel comfortable with how to use sort_values() to do fairly robust tasks quickly.

# Cleaning the DataFrame

## Covered in this Section

1) How to Drop a Column in the DataFrame<br>
2) How to Remove Rows that have NaN in any Column<br>
3) How to Remove Rows that have NaN in a Specific Column<br>
4) How to Convert DataFrame Data Types (from Float to Int)

## How to Drop a Column in Pandas DataFrame

In [1]:
import pandas as pd
df = pd.read_csv("data/titanic.csv")
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

With our data loaded, let's go ahead and jump right into the chapter. Imagine that we have a large DataFrame, but we are not interested in a couple columns. This is especially import when your DataFrame has 10s or 100s of columns. In these instances, you need to examine the DataFrame without the useless data. Imagine that we wanted to study the Titanic data but knew that Parch and Ticket were categories that we did not need. We can use df.drop() to pass an argument to remove those specific columns.

In [2]:
df.drop(columns=["Parch", "Ticket"])

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

        Fare Cabin Embarked  
0     7.2500   NaN        S  
1    71.2833   C85        C  
2     7.9250   NaN        S  
3    53.1000  C123        S  
4     8.0500   NaN        S  
..       ...   ...      ...  
886  13.0000   NaN        S  
887  30.0000   B42        S  
888  23.4500   NaN        S  
889  30.0000  C148        C  
890   7.7500   NaN        Q  

[891 rows x 10 columns]

## How to Remove Rows that have NaN in any Column

One of the biggest problems in datasets is the absence of data. If you are training a machine learning model or just performing quantitative analysis, rows that have missing values, or NaN, can radically alter your results. It is often good practice to ignore that data or alter it in some way. Let's presume that we want to simply remove it from our dataset. To do that, we can use df.dropna() which will remove all rows that have any instance of NaN in any column.

In [3]:
df.dropna()

PassengerId  Survived  Pclass  \
1              2         1       1   
3              4         1       1   
6              7         0       1   
10            11         1       3   
11            12         1       1   
..           ...       ...     ...   
871          872         1       1   
872          873         0       1   
879          880         1       1   
887          888         1       1   
889          890         1       1   

                                                  Name     Sex   Age  SibSp  \
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
6                              McCarthy, Mr. Timothy J    male  54.0      0   
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
11                            Bonnell, Miss. Elizabeth  female  58.0      0   
..                                                 ...     ...   ...    ...   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
872                           Carlsson, Mr. Frans Olof    male  33.0      0   
879      Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   

     Parch    Ticket     Fare        Cabin Embarked  
1        0  PC 17599  71.2833          C85        C  
3        0    113803  53.1000         C123        S  
6        0     17463  51.8625          E46        S  
10       1   PP 9549  16.7000           G6        S  
11       0    113783  26.5500         C103        S  
..     ...       ...      ...          ...      ...  
871      1     11751  52.5542          D35        S  
872      0       695   5.0000  B51 B53 B55        S  
879      1     11767  83.1583          C50        C  
887      0    112053  30.0000          B42        S  
889      0    111369  30.0000         C148        C  

[183 rows x 12 columns]

## How to Remove Rows that have NaN in a Specific Column

In some instances, though, we don't want to remove an entire row just because of NaN in one column. Maybe that column is not as important for quantitative analysis and we are not planning to include it in our analysis, but we still want to see it. A good example of this is the column Cabin which is a string or Age which is a float (we'll get to that in a moment). Let's say we want to remove all rows that have NaN in the Age column. We can use the command below.

In [7]:
df2 = df[df["Age"].notna()]

In [8]:
df2

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female  39.0      0   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
885      5            382652  29.1250   NaN        Q  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[714 rows x 12 columns]

As we can see, the size of our DataFrame dropped from 891 rows to 714.

## How to Convert DataFrame Data Types (from Float to Int)

In other instances, it may be important not to simply remove a column, but alter it into a different type of data. In this dataset, Age is a float. This is to account for infants who were below the age of 1 on the Titanic. Let's presume that we want to convert all these floats to integers. To do that we can use the .astype() method on a specific row.

In [11]:
df2.Age = df2.Age.astype(int)

In [12]:
df2

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
885          886         0       3   
886          887         0       2   
887          888         1       1   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex  Age  SibSp  \
0                              Braund, Mr. Owen Harris    male   22      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female   38      1   
2                               Heikkinen, Miss. Laina  female   26      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   35      1   
4                             Allen, Mr. William Henry    male   35      0   
..                                                 ...     ...  ...    ...   
885               Rice, Mrs. William (Margaret Norton)  female   39      0   
886                              Montvila, Rev. Juozas    male   27      0   
887                       Graham, Miss. Margaret Edith  female   19      0   
889                              Behr, Mr. Karl Howell    male   26      0   
890                                Dooley, Mr. Patrick    male   32      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
885      5            382652  29.1250   NaN        Q  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[714 rows x 12 columns]

Now our Age column is no longer a float, rather an integer.